### PR TITLE
[Dynamic Instrumentation] Fix stability issues

### PR DESCRIPTION
**Pull request by GreenMatan**
### What does this PR do?
* Fixed few nil dereferences that caused crashes ([DEBUG-3454](https://datadoghq.atlassian.net/jira/software/c/projects/DEBUG/boards/760?assignee=6172d61db9c549006f9adf14&selectedIssue=DEBUG-3454))
* Fixed a race condition between map reading & writing, resulting in a crash ([DEBUG-3211](https://datadoghq.atlassian.net/jira/software/c/projects/DEBUG/boards/760?assignee=6172d61db9c549006f9adf14&selectedIssue=DEBUG-3211))
* One probe used to sabotage all other probes in a service if there was a failure during DWARF inspection ([DEBUG-3230](https://datadoghq.atlassian.net/jira/software/c/projects/DEBUG/boards/760?assignee=6172d61db9c549006f9adf14&selectedIssue=DEBUG-3230))
* Similar to the previous point, one process used to fail all other processes inspection if there was an error while processing probes for particular process ([DEBUG-3205](https://datadoghq.atlassian.net/jira/software/c/projects/DEBUG/boards/760?assignee=6172d61db9c549006f9adf14&selectedIssue=DEBUG-3205))

### Motivation

Improve stability of Go DI

### Describe how you validated your changes
I ran locally our e2e tests & exploration tests and verified the proposed changes improved our stability.

[DEBUG-3454]: https://datadoghq.atlassian.net/browse/DEBUG-3454?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ
[DEBUG-3230]: https://datadoghq.atlassian.net/browse/DEBUG-3230?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ
[DEBUG-3205]: https://datadoghq.atlassian.net/browse/DEBUG-3205?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ
[DEBUG-3211]: https://datadoghq.atlassian.net/browse/DEBUG-3211?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

[DEBUG-3124]: https://datadoghq.atlassian.net/browse/DEBUG-3124?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: pkg/dynamicinstrumentation/testutil/exploration_e2e_test.go [hunk 1]
```diff
@@ -0,0 +1,1206 @@
+// Unless explicitly stated otherwise all files in this repository are licensed
+// under the Apache License Version 2.0.
+// This product includes software developed at Datadog (https://www.datadoghq.com/).
+// Copyright 2016-present Datadog, Inc.
+
+//go:build exploration_test_linux_bpf
```
Comment (GreenMatan) [inline pkg/dynamicinstrumentation/testutil/exploration_e2e_test.go:6]: Deliberately avoid being part of the build process. The exploration testing, for now, will be a tool we use locally. In future PR it will be part of the CI pipeline and kick when di related code changes

**Comment by grantseltzer** [comment]
I can't get the tests to actually run. Issues with running git:

```
=== RUN   TestIntegration
download /tmp/protobuf-integration-1060272402/src/google.golang.org/protobuf/.cache/protobuf-27.0
    integration_test.go:267: executing (git clone https://github.com/protocolbuffers/protobuf protobuf-27.0): exit status 128
        Cloning into 'protobuf-27.0'...
        error: RPC failed; curl 92 HTTP/2 stream 0 was not closed cleanly: CANCEL (err 8)
        error: 4217 bytes of body are still expected
        fetch-pack: unexpected disconnect while reading sideband packet
        fatal: early EOF
        fatal: fetch-pack: invalid index-pack output
--- FAIL: TestIntegration (118.60s)
FAIL
FAIL	google.golang.org/protobuf	120.028s
FAIL
    exploration_e2e_test.go:1038:
        	Error Trace:	/home/vagrant/datadog-agent/pkg/dynamicinstrumentation/testutil/exploration_e2e_test.go:1038
        	Error:      	Received unexpected error:
        	            	exit status 1
        	Test:       	TestExplorationGoDI
--- FAIL: TestExplorationGoDI (142.58s)
FAIL
exit status 1
FAIL	github.com/DataDog/datadog-agent/pkg/dynamicinstrumentation/testutil	142.615s
```

Regardless I think dependencies would be better served as git submodules, what do you think?.

File: pkg/dynamicinstrumentation/testutil/exploration_e2e_test.go [hunk 2]
```diff
@@ -0,0 +1,1206 @@
...
+		return fmt.Errorf("binary analysis failed: %v", err)
+	}
+	pt.mu.Lock()
+	info.State = StateRunning
+	pt.mu.Unlock()
+	return nil
+}
+
+func (pt *ProcessTracker) scanProcessTree() error {
+	if err := syscall.Kill(-g_cmd.Process.Pid, syscall.SIGSTOP); err != nil {
```
Comment (grantseltzer) [inline pkg/dynamicinstrumentation/testutil/exploration_e2e_test.go:647]: Can you document more the need to use signals for stopping/continuing procs?

My understanding so far is that we download the protobuf library, compile all test binaries, attach probes to all functions in the binaries, then run them and inspect results. Why is it necessary to start a process group and orchestrate with signals instead of running the binaries one at a time?

**Comment by grantseltzer** [comment]
1. Let's split this into two PRs, one for stability fixes and another for exploration testing.

1. Generally speaking i'd like to discuss making the testing set up simpler. Maybe we can start by having a service we own (or build on top of the existing sample service) which can use protobuf. We could even start multiple instances of it with a runner for the sake of stress testing. Each instance can even use multiple go routines for portions of the test. It would make it a lot easier to know what paths are actually called, and what compare event output we expect for accuracy.

File: pkg/dynamicinstrumentation/testutil/exploration_e2e_test.go (hunk 2 above)
Comment (GreenMatan) [reply pkg/dynamicinstrumentation/testutil/exploration_e2e_test.go:647]: I'm not sure I follow. How could you attach probes to all functions in a binaries - without spawning up a process? As far as I understand it, we have to have a living process in order to attach uprobes.
I liked more the approach of running the tests in a blackbox without caring much about their nature in order to have an infra we could easily add more projects to test. I didn't want to create a test that manually puts probe on predefined locations, and rather do it automatically (reading functions from the binary and putting probes).
I do agree, though, that we have a challenge in determining which functions are actually executing. I don't think this challenge rule out the whole technique just yet. I could tackle this challenge by either statically analyzing the code paths of the tests and put probes on the symbols we find or profile protobuf as the first step, extract bunch of symbols and then rerun the test with exploration testing and put probes on those functions. Both of this techniques do not change the whole infra I've built but rather completes it.

Having said all of that, I think the usefulness of the exploration tests as they stand right now is still huge. I managed to find 6+ crashes due to nil/invalid indexing dereferences, 1 race condition and 2 logical bugs by simply running them. Debugging it was also a bless.
I don't say that exploration tests are a complete task as they currently stand, but they're totally a tool to assess the readiness to private beta.

Comment (GreenMatan) [reply pkg/dynamicinstrumentation/testutil/exploration_e2e_test.go:647]: I considered a couple of approaches and ended up going in the approach of running the tests in a blackbox without caring much about their nature. The infra is pretty flexible and can be easily adjusted to add more projects. I didn't want to manually import the project, look into how to run it continuously and be forced to put probes on predefined locations.
I do agree, though, that we have a challenge in determining which functions are actually executing. I don't think this challenge rule out the whole technique just yet. I can deal this challenge by either statically analyzing the code paths of the tests and put probes on the symbols we find or profile protobuf's tests, extract bunch of symbols and then rerun the test with exploration testing and put probes on those symbols. Both of this techniques do not change the whole infra I've built but rather completes it.

Having said all of that, I still believe the exploration tests are incredibly useful in their current state. Simply running them helped me uncover 6+ crashes caused by nil/invalid indexing dereferences, 1 race condition, and 2 logical bugs - bugs that would have otherwise been really hard to find. Debugging it was also a bless. The fact that numerous subtests involve in the execution of protobuf, some of which are very short-lived, also helped uncover these edge cases.
I don't say that exploration tests are a complete task as they currently stand, but they're totally a tool to assess the readiness to private beta. We could choose to create more tests either by utilizing this infra or creating a new one, but the current exploration tests did prove themselves.

**Comment by GreenMatan** [comment]
I've splitted the PR into Stability & Exploration Testing. The exploration testing PR is: #34423 and is stacked on top of this PR.

File: pkg/dynamicinstrumentation/diconfig/config_manager.go [hunk 3]
```diff
@@ -243,14 +243,17 @@ func (cm *RCConfigManager) readConfigs(r *ringbuf.Reader, procInfo *ditypes.Proc
...
-	if err != nil {
-		log.Errorf("couldn't inspect binary: %v\n", err)
-		return
+
+	if procInfo.TypeMap == nil {
+		err := AnalyzeBinary(procInfo)
+		if err != nil {
+			log.Errorf("couldn't inspect binary: %v\n", err)
+			return
+		}
```
Comment (grantseltzer) [inline pkg/dynamicinstrumentation/diconfig/config_manager.go:252]: There's other fields which are set in procInfo from AnalyzeBinary besides the type map, so i'm not sure if this is the correct way to do this check. 

I originally thought that it would be an issue because when the binary is inspected we're only analyzing types for target functions, so if a config is updated, we wouldn't be able to parse types for the new configuration. However testing this branch doesn't show any issues.

What were the reservations you had about this particular change?

Comment (grantseltzer) [reply pkg/dynamicinstrumentation/diconfig/config_manager.go:252]: Ok I did further investigation and realized this code doesn't belong here at all. `AnalyzeBinary` needs to be called for the sake of the RC config manager, but the memory config manager (i.e. what file based configuration is built on) has separate callbacks for when procs/configs are updated which call AnalyzeBinary already.

So remove this call to AnalyzeBinary all together and add it before the call to `applyConfigUpdate()` on line 239 (without the condition nil check)

Comment (GreenMatan) [reply pkg/dynamicinstrumentation/diconfig/config_manager.go]: Done in c521706bedf6987961f29d07701111a8b3cac9ec.

File: pkg/dynamicinstrumentation/ditypes/config.go [hunk 4]
```diff
@@ -172,12 +174,16 @@ func (pi *ProcessInfo) CloseConfigUprobe() error {
...
 func (pi *ProcessInfo) SetUprobeLink(probeID ProbeID, l *link.Link) {
+	pi.mu.Lock()
+	defer pi.mu.Unlock()
 	pi.InstrumentationUprobes[probeID] = l
 }
 
 // CloseUprobeLink closes the probe and deletes the link for the probe
 // in the tracked process
 func (pi *ProcessInfo) CloseUprobeLink(probeID ProbeID) error {
+	pi.mu.Lock()
```
Comment (tylfin) [inline pkg/dynamicinstrumentation/ditypes/config.go:185]: We could probably utilize a [`sync.Map`](https://pkg.go.dev/sync#Map) here to ensure `InstrumentationUprobes` is thread-safe. Does the use-case fit the criteria? 

> The sync.Map type is optimized for two common use cases: (1) when the entry for a given key is only ever written once but read many times, as in caches that only grow, or (2) when multiple goroutines read, write, and overwrite entries for disjoint sets of keys

**Comment by grantseltzer** [comment]
@GreenMatan You need to sign CLA to get tests to pass.

File: pkg/dynamicinstrumentation/ditypes/config.go (hunk 4 above)
Comment (GreenMatan) [reply pkg/dynamicinstrumentation/ditypes/config.go:188]: I see the value in using `sync.Map` for simplicity and broader thread-safety beyond this source file. I’ve noted this for future consideration but opted for a simple mutex since adding/removing probes in InstrumentationUprobes is infrequent.
In the meantime, I’ve ensured additional protections in 08bbb5ca0e3b60bb114299e210867816b8274a4f.

**Review by tylfin — APPROVED** [review]
